src/mongocxx/test/spec/monitoring.cpp

```diff
@@ -46,38 +75,71 @@ void apm_checker::compare(bsoncxx::array::view expectations,
     _events.erase(std::remove_if(_events.begin(), _events.end(), is_ignored), std::end(_events));
     for (auto expectation : expectations) {
         auto expected = expectation.get_document().view();
-
-        if (events_iter == _events.end()) {
-            FAIL("Out of events, expected to find event " << to_json(expected) << "\n\n");
-            print_all();
-            REQUIRE(false);
-        }
-
-        CAPTURE(to_json(*events_iter), expectation);
+        REQUIRE(events_iter != _events.end());
+        CAPTURE(print_all());
         REQUIRE_BSON_MATCHES_V(*events_iter, expected, match_visitor);
         events_iter++;
     }
 
-    REQUIRE((allow_extra || events_iter == _events.end()));
+    if (!allow_extra)
+        REQUIRE(events_iter == _events.end());
 }
 
 void apm_checker::has(bsoncxx::array::view expectations) {
     for (auto expectation : expectations) {
         auto expected = expectation.get_document().view();
-        CAPTURE(to_json(expected).c_str());
+        CAPTURE(to_json(expected).c_str(), print_all());
         REQUIRE(std::find_if(_events.begin(), _events.end(), [&](bsoncxx::document::view doc) {
                     return test_util::matches(doc, expected);
                 }) != _events.end());
     }
 }
 
-void apm_checker::print_all() {
-    printf("\n\n");
-    printf("APM Checker contents: \n");
+std::string apm_checker::print_all() {
```

Comment:
This function used `printf`, which causes issues with `CATCH2`. Since that framework maintains its own stack for output, this is always printed out of order, making it hard to find and piece together.

Reply:
Great, that level of detail will probably save countless hours.